bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/ReviewDataSourceServlet.java

```diff
@@ -0,0 +1,134 @@
+import com.adobe.granite.ui.components.ds.DataSource;
+import com.adobe.granite.ui.components.ds.SimpleDataSource;
+import com.adobe.granite.ui.components.ds.ValueMapResource;
+
+@Component(
+    service = { Servlet.class },
+    property = {
+        "sling.servlet.resourceTypes=" + FormConstants.RT_FD_FORM_REVIEW_DATASOURCE_V1,
+        "sling.servlet.methods=GET",
+        "sling.servlet.extensions=html"
+    })
+public class ReviewDataSourceServlet extends AbstractDataSourceServlet {
+
+    /**
+     * Defines the form meta data type. Possible values: {@code submitAction},
+     * {@code prefillServiceProvider}
+     *
+     * @todo: Add other metadata types here like fragment, actions etc
+     */
+    @Reference
+    private transient ExpressionResolver expressionResolver;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReviewDataSourceServlet.class);
+
+    @NotNull
+    @Override
+    protected ExpressionResolver getExpressionResolver() {
+        return expressionResolver;
+    }
+
+    @Override
+    protected void doGet(@NotNull SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response) {
+
+        ResourceResolver resourceResolver = request.getResourceResolver();
+        String componentInstancePath = request.getRequestPathInfo().getSuffix();
+        List<Resource> resources = new ArrayList<>();
+        if (resourceResolver != null) {
+            Resource componentInstance = resourceResolver.getResource(componentInstancePath);
+            Resource formInstance = ComponentUtils.getFormContainer(componentInstance);
+            FormContainer formContainer = formInstance.adaptTo(FormContainer.class);
+            List<Base> panels = (List<Base>) getPanels(formContainer);
+            for (Base panel : panels) {
+                String id = panel != null ? panel.getId() : "";
+                String title = "";
+                if (panel != null) {
+                    Label label = panel.getLabel();
+                    if (label != null) {
+                        String value = label.getValue();
+                        if (value != null) {
+                            title = value;
+                        }
+                    }
+                }
+                if (id != null && title != null) {
+                    resources.add(getResourceForDropdownDisplay(resourceResolver, title, id));
+                } else {
+                    LOGGER.error("[AF] id or title is null");
+                }
+            }
+        }
+        SimpleDataSource actionTypeDataSource = new SimpleDataSource(resources.iterator());
+        request.setAttribute(DataSource.class.getName(), actionTypeDataSource);
+    }
+
+    List<? extends ComponentExporter> getPanels(FormComponent formContainer) {
+        if (formContainer instanceof FormContainer) {
+            while (formContainer instanceof Container && ((Container) formContainer).getItems().size() == 1) {
```

Comment:
FormContainer extends Container. If an object is instance of Child class, it will also be instance of Parent class.
You can remove this condition: formContainer instanceof Container